# Breast Cancer Analysis Prediction
####OBJECTIVE OF THE PROJECT:
  * Cancer is one of the leading causes of deaths today. While developing therapeutic approaches is key, the role of correct classification cannot be     undermined.
  * For this reason, we are going to analyze the Breast Cancer Wisconsin (Diagnostic) DataSet, obtained from Kaggle. By applying machine learning algorithms, we are going to predict the Stage of Breast Cancer M (Malignant) and B (Bengin).
  
#####Author:
* Anushree

In [2]:
displayHTML("<img src ='https://sites.psu.edu/breastcancerawarenesshub/files/2016/09/breast_cancer_stats-2k17b5s.png'>")

## Collecting data
######Data lineage:  
* This data originated from The UCI Machine Learning Repository. (Find it here :https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29). 
* This data has been used and cited in multiple papers for research related machine learning techniques and algorithms.
* We are using the compiled .csv version of this data which has been put up on Kaggle by UCI for Predictive analysis.

In [4]:
%sh
wget -O breastcancer https://www.dropbox.com/s/qc1nabe8vk7642t/data.csv?dl=0

--2019-12-14 04:48:13-- https://www.dropbox.com/s/qc1nabe8vk7642t/data.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.1.1, 2620:100:6016:1::a27d:101
Connecting to www.dropbox.com (www.dropbox.com)|162.125.1.1|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/qc1nabe8vk7642t/data.csv [following]
--2019-12-14 04:48:13-- https://www.dropbox.com/s/raw/qc1nabe8vk7642t/data.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://uc63fc1cb87736f6dbd210024913.dl.dropboxusercontent.com/cd/0/inline/AuNSimvmLc30WUFCdgkt_dL5IddKCSiNhYOO27BXJydCwCyD-CeNGJEwXr0gX-OaZEsmmDzV7i39u8QgIgLvug1aNL_zliZgi9oJiq86FeL24TTcUG9OUHfPUvHsykb6Ecc/file# [following]
--2019-12-14 04:48:14-- https://uc63fc1cb87736f6dbd210024913.dl.dropboxusercontent.com/cd/0/inline/AuNSimvmLc30WUFCdgkt_dL5IddKCSiNhYOO27BXJydCwCyD-CeNGJEwXr0gX-OaZEsmmDzV7i39u8QgIgLvug1aNL_zliZgi9oJiq86FeL24TTcUG9OUHfPUvHsykb6Ecc/file
Resolving uc63fc1cb87736f6dbd210024913.dl.dropboxusercontent.com (uc63fc1cb87736f6dbd210024913.dl.dropboxusercontent.com)... 162.125.1.6, 2620:100:6016:6::a27d:106
Connecting to uc63fc1cb87736f6dbd210024913.dl.dropboxusercontent.com (uc63fc1cb87736f6dbd210024913.dl.dropboxusercontent.com)|162.125.1.6|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 125204 (122K) [text/plain]
Saving to: ‘breastcancer’

 0K .......... .......... .......... .......... .......... 40% 3.46M 0s
 50K .......... .......... .......... .......... .......... 81% 12.5M 0s
 100K .......... .......... .. 100% 21.5M=0.02s

2019-12-14 04:48:14 (6.27 MB/s) - ‘breastcancer’ saved [125204/125204]

In [5]:
%sh
ls /databricks/driver         

breastcancer
conf
derby.log
eventlogs
ganglia
logs

In [6]:
df = spark.read.csv('file:///databricks/driver/breastcancer',inferSchema='true',header='true')

####DATA EXPLAINATION:
  * This dataset contains the diagnosis and a set of 30 features describing the characteristics of the cell nuclei present in the digitized image of a of a fine needle aspirate (FNA) of a breast mass.
  
  ####Attribute Information:

1) ID number 2) Diagnosis (M = malignant, B = benign) 

  #####Ten real-valued features are computed for each cell nucleus:
  * radius (mean of distances from center to points on the perimeter);
  * texture (standard deviation of gray-scale values);
  * perimeter;
  * area;
  * smoothness (local variation in radius lengths);
  * compactness (perimeter^2 / area - 1.0);
  * concavity (severity of concave portions of the contour);
  * concave points (number of concave portions of the contour);
  * symmetry;
  * fractal dimension (“coastline approximation” - 1).

In [8]:
df.printSchema()

root
-- id: integer (nullable = true)
-- diagnosis: string (nullable = true)
-- radius_mean: double (nullable = true)
-- texture_mean: double (nullable = true)
-- perimeter_mean: double (nullable = true)
-- area_mean: double (nullable = true)
-- smoothness_mean: double (nullable = true)
-- compactness_mean: double (nullable = true)
-- concavity_mean: double (nullable = true)
-- concave points_mean: double (nullable = true)
-- symmetry_mean: double (nullable = true)
-- fractal_dimension_mean: double (nullable = true)
-- radius_se: double (nullable = true)
-- texture_se: double (nullable = true)
-- perimeter_se: double (nullable = true)
-- area_se: double (nullable = true)
-- smoothness_se: double (nullable = true)
-- compactness_se: double (nullable = true)
-- concavity_se: double (nullable = true)
-- concave points_se: double (nullable = true)
-- symmetry_se: double (nullable = true)
-- fractal_dimension_se: double (nullable = true)
-- radius_worst: double (nullable = true)
-- texture_worst: double (nullable = true)
-- perimeter_worst: double (nullable = true)
-- area_worst: double (nullable = true)
-- smoothness_worst: double (nullable = true)
-- compactness_worst: double (nullable = true)
-- concavity_worst: double (nullable = true)
-- concave points_worst: double (nullable = true)
-- symmetry_worst: double (nullable = true)
-- fractal_dimension_worst: double (nullable = true)
-- _c32: string (nullable = true)

In [9]:
display(df)

id diagnosis radius_mean texture_mean perimeter_mean area_mean smoothness_mean compactness_mean concavity_mean concave points_mean symmetry_mean fractal_dimension_mean radius_se texture_se perimeter_se area_se smoothness_se compactness_se concavity_se concave points_se symmetry_se fractal_dimension_se radius_worst texture_worst perimeter_worst area_worst smoothness_worst compactness_worst concavity_worst concave points_worst symmetry_worst fractal_dimension_worst _c32 842302 M 17.99 10.38 122.8 1001.0 0.1184 0.2776 0.3001 0.1471 0.2419 0.07871 1.095 0.9053 8.589 153.4 0.006399 0.04904 0.05373 0.01587 0.03003 0.006193 25.38 17.33 184.6 2019.0 0.1622 0.6656 0.7119 0.2654 0.4601 0.1189 null 842517 M 20.57 17.77 132.9 1326.0 0.08474 0.07864 0.0869 0.07017 0.1812 0.05667 0.5435 0.7339 3.398 74.08 0.005225 0.01308 0.0186 0.0134 0.01389 0.003532 24.99 23.41 158.8 1956.0 0.1238 0.1866 0.2416 0.186 0.275 0.08902 null 84300903 M 19.69 21.25 130.0 1203.0 0.1096 0.1599 0.1974 0.1279 0.2069 0.05999 0.7456 0.7869 4.585 94.03 0.00615 0.04006 0.03832 0.02058 0.0225 0.004571 23.57 25.53 152.5 1709.0 0.1444 0.4245 0.4504 0.243 0.3613 0.08758 null 84348301 M 11.42 20.38 77.58 386.1 0.1425 0.2839 0.2414 0.1052 0.2597 0.09744 0.4956 1.156 3.445 27.23 0.00911 0.07458 0.05661 0.01867 0.05963 0.009208 14.91 26.5 98.87 567.7 0.2098 0.8663 0.6869 0.2575 0.6638 0.173 null 84358402 M 20.29 14.34 135.1 1297.0 0.1003 0.1328 0.198 0.1043 0.1809 0.05883 0.7572 0.7813 5.438 94.44 0.01149 0.02461 0.05688 0.01885 0.01756 0.005115 22.54 16.67 152.2 1575.0 0.1374 0.205 0.4 0.1625 0.2364 0.07678 null 843786 M 12.45 15.7 82.57 477.1 0.1278 0.17 0.1578 0.08089 0.2087 0.07613 0.3345 0.8902 2.217 27.19 0.00751 0.03345 0.03672 0.01137 0.02165 0.005082 15.47 23.75 103.4 741.6 0.1791 0.5249 0.5355 0.1741 0.3985 0.1244 null 844359 M 18.25 19.98 119.6 1040.0 0.09463 0.109 0.1127 0.074 0.1794 0.05742 0.4467 0.7732 3.18 53.91 0.004314 0.01382 0.02254 0.01039 0.01369 0.002179 22.88 27.66 153.2 1606.0 0.1442 0.2576 0.3784 0.1932 0.3063 0.08368 null 84458202 M 13.71 20.83 90.2 577.9 0.1189 0.1645 0.09366 0.05985 0.2196 0.07451 0.5835 1.377 3.856 50.96 0.008805 0.03029 0.02488 0.01448 0.01486 0.005412 17.06 28.14 110.6 897.0 0.1654 0.3682 0.2678 0.1556 0.3196 0.1151 null 844981 M 13.0 21.82 87.5 519.8 0.1273 0.1932 0.1859 0.09353 0.235 0.07389 0.3063 1.002 2.406 24.32 0.005731 0.03502 0.03553 0.01226 0.02143 0.003749 15.49 30.73 106.2 739.3 0.1703 0.5401 0.539 0.206 0.4378 0.1072 null 84501001 M 12.46 24.04 83.97 475.9 0.1186 0.2396 0.2273 0.08543 0.203 0.08243 0.2976 1.599 2.039 23.94 0.007149 0.07217 0.07743 0.01432 0.01789 0.01008 15.09 40.68 97.65 711.4 0.1853 1.058 1.105 0.221 0.4366 0.2075 null 845636 M 16.02 23.24 102.7 797.8 0.08206 0.06669 0.03299 0.03323 0.1528 0.05697 0.3795 1.187 2.466 40.51 0.004029 0.009269 0.01101 0.007591 0.0146 0.003042 19.19 33.88 123.8 1150.0 0.1181 0.1551 0.1459 0.09975 0.2948 0.08452 null 84610002 M 15.78 17.89 103.6 781.0 0.0971 0.1292 0.09954 0.06606 0.1842 0.06082 0.5058 0.9849 3.564 54.16 0.005771 0.04061 0.02791 0.01282 0.02008 0.004144 20.42 27.28 136.5 1299.0 0.1396 0.5609 0.3965 0.181 0.3792 0.1048 null 846226 M 19.17 24.8 132.4 1123.0 0.0974 0.2458 0.2065 0.1118 0.2397 0.078 0.9555 3.568 11.07 116.2 0.003139 0.08297 0.0889 0.0409 0.04484 0.01284 20.96 29.94 151.7 1332.0 0.1037 0.3903 0.3639 0.1767 0.3176 0.1023 null 846381 M 15.85 23.95 103.7 782.7 0.08401 0.1002 0.09938 0.05364 0.1847 0.05338 0.4033 1.078 2.903 36.58 0.009769 0.03126 0.05051 0.01992 0.02981 0.003002 16.84 27.66 112.0 876.5 0.1131 0.1924 0.2322 0.1119 0.2809 0.06287 null 84667401 M 13.73 22.61 93.6 578.3 0.1131 0.2293 0.2128 0.08025 0.2069 0.07682 0.2121 1.169 2.061 19.21 0.006429 0.05936 0.05501 0.01628 0.01961 0.008093 15.03 32.01 108.8 697.7 0.1651 0.7725 0.6943 0.2208 0.3596 0.1431 null 84799002 M 14.54 27.54 96.73 658.8 0.1139 0.1595 0.1639 0.07364 0.2303 0.07077 0.37 1.033 2.879 32.55 0.005607 0.0424 0.04741 0.0109 0.01857 0.005466 17.46 37.13 124.1 943.2 0.16

In [10]:
df.count()

Out[39]: 569

In [11]:
#counting number of rows in dataframe
len(df.columns)

Out[40]: 33

In [12]:
df1=df.summary()
display(df1) #we can ignore the summary for the id and diagnosis and _c32 column. Only the count summary for those items can be taken into consideration

summary id diagnosis radius_mean texture_mean perimeter_mean area_mean smoothness_mean compactness_mean concavity_mean concave points_mean symmetry_mean fractal_dimension_mean radius_se texture_se perimeter_se area_se smoothness_se compactness_se concavity_se concave points_se symmetry_se fractal_dimension_se radius_worst texture_worst perimeter_worst area_worst smoothness_worst compactness_worst concavity_worst concave points_worst symmetry_worst fractal_dimension_worst _c32 count 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 569 0 mean 3.0371831432337433E7 null 14.127291739894563 19.28964850615117 91.96903339191566 654.8891036906857 0.096360281195079 0.10434098418277686 0.08879931581722322 0.048919145869947236 0.181161862917399 0.06279760984182778 0.4051720562390161 1.2168534270650269 2.8660592267135288 40.33707908611603 0.007040978910369071 0.025478138840070306 0.031893716344463946 0.011796137082601056 0.020542298769771532 0.0037949038664323383 16.269189806678394 25.677223198594014 107.2612126537786 880.5831282952545 0.13236859402460469 0.25426504393673144 0.27218848330404205 0.11460622319859404 0.29007557117750454 0.08394581722319855 null stddev 1.2502058561222367E8 null 3.5240488262120793 4.301035768166948 24.2989810387549 351.9141291816529 0.014064128137673616 0.0528127579325122 0.07971980870789354 0.03880284485915361 0.027414281336035712 0.007060362795084457 0.2773127329861039 0.5516483926172022 2.0218545540421085 45.49100551613178 0.003002517943839066 0.01790817932567739 0.030186060322988394 0.006170285174046867 0.008266371528798402 0.002646070967089195 4.833241580469323 6.1462576230383235 33.602542269036356 569.3569926699494 0.022832429404835465 0.157336488913742 0.20862428060813235 0.06573234119594208 0.06186746753751871 0.01806126734889399 null min 8670 B 6.981 9.71 43.79 143.5 0.05263 0.01938 0.0 0.0 0.106 0.04996 0.1115 0.3602 0.757 6.802 0.001713 0.002252 0.0 0.0 0.007882 8.948E-4 7.93 12.02 50.41 185.2 0.07117 0.02729 0.0 0.0 0.1565 0.05504 null 25% 869218 null 11.7 16.17 75.17 420.3 0.08637 0.06492 0.02956 0.02031 0.1619 0.0577 0.2324 0.8339 1.606 17.85 0.005169 0.01308 0.01509 0.007638 0.01516 0.002248 13.01 21.08 84.11 515.3 0.1166 0.1472 0.1145 0.06493 0.2504 0.07146 null 50% 906024 null 13.37 18.84 86.24 551.1 0.09587 0.09263 0.06154 0.0335 0.1792 0.06154 0.3242 1.108 2.287 24.53 0.00638 0.02045 0.02589 0.01093 0.01873 0.003187 14.97 25.41 97.66 686.5 0.1313 0.2119 0.2267 0.09993 0.2822 0.08004 null 75% 8813129 null 15.78 21.8 104.1 782.7 0.1053 0.1304 0.1307 0.074 0.1957 0.06612 0.4789 1.474 3.357 45.19 0.008146 0.03245 0.04205 0.01471 0.02348 0.004558 18.79 29.72 125.4 1084.0 0.146 0.3391 0.3829 0.1614 0.3179 0.09208 null max 911320502 M 28.11 39.28 188.5 2501.0 0.1634 0.3454 0.4268 0.2012 0.304 0.09744 2.873 4.885 21.98 542.2 0.03113 0.1354 0.396 0.05279 0.07895 0.02984 36.04 49.54 251.2 4254.0 0.2226 1.058 1.252 0.291 0.6638 0.2075 null

## Data Visualtization

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt   #Data visualisation libraries 
import seaborn as sns  # https://seaborn.pydata.org/generated/seaborn.barplot.html
import sklearn as sk
import sklearn.model_selection
import sklearn.linear_model
from sklearn import preprocessing

In [15]:
fig, ax = plt.subplots()
fig.set_size_inches(14, 14)
corr = pdata.corr()
sns.heatmap(corr,annot=True, linewidths=.5, fmt= '.1f',ax=ax)
display(fig)

######The result from the Heat map shows us that:
* The radius_mean, perimeter_mean and area_mean are correlated with each other.
* Compactness_mean, concavity_mean and concave points_mean are correlated with each other.
* Also radius_se, perimeter_se and area_se are correlated.

In [17]:
pdata = df.toPandas()
sns.set(style="white", font_scale=.6)
fig, ax = plt.subplots()
fig.set_size_inches( 4, 4)
g = sns.distplot(pdata['radius_mean'])
display(fig)

In [18]:
pdata = df.toPandas()
sns.set(style="white", font_scale=.6)
fig, ax = plt.subplots()
fig.set_size_inches( 4, 4)
g = sns.distplot(pdata['area_mean'])
display(fig)

In [19]:
pdata = df.toPandas()
sns.set(style="white", font_scale=.6)
fig, ax = plt.subplots()
fig.set_size_inches( 4, 4)
g = sns.distplot(pdata['concavity_mean'])
display(fig)

In [20]:
pdata = df.toPandas()
sns.set(style="white", font_scale=.6)
fig, ax = plt.subplots()
fig.set_size_inches( 4, 4)
g = sns.distplot(pdata['radius_se'])
display(fig)

In [21]:
pdata = df.toPandas()
sns.set(style="white", font_scale=.6)
fig, ax = plt.subplots()
fig.set_size_inches( 4, 4)
g = sns.distplot(pdata['area_se'])
display(fig)

In [22]:
pdata = df.toPandas()
sns.set(style="white", font_scale=.6)
fig, ax = plt.subplots()
fig.set_size_inches( 4, 4)
g = sns.distplot(pdata['concavity_se'])
display(fig)

From the above histograms we can see radius_mean, area_mean, concavity_mean show more difference between malignant and benign populations compared to others.

In [24]:
pdata['diagnosis']= pdata['diagnosis'].map({'M':1,'B':0})

In [25]:
display(sns.countplot( pdata['diagnosis'],label="Count"))

In [26]:
features_mean=list(pdata.columns[1:11])
# split dataframe into two based on diagnosis
dfM=pdata[pdata['diagnosis'] ==1]
dfB=pdata[pdata['diagnosis'] ==0]

In [27]:
plt.rcParams.update({'font.size': 8})
fig, axes = plt.subplots(nrows=5, ncols=2, figsize=(8,10))
axes = axes.ravel()
for idx,ax in enumerate(axes):
    ax.figure
    binwidth= (max(pdata[features_mean[idx]]) - min(pdata[features_mean[idx]]))/50
    ax.hist([dfM[features_mean[idx]],dfB[features_mean[idx]]], bins=np.arange(min(pdata[features_mean[idx]]), max(pdata[features_mean[idx]]) + binwidth, binwidth) , alpha=0.5,stacked=True, density = True, label=['M','B'],color=['r','g'])
    ax.legend(loc='upper right')
    ax.set_title(features_mean[idx])
fig.set_size_inches(10, 10)
fig =plt.tight_layout()
display(fig)

##ADDITIONAL VISUALIZATIONS

In [29]:
y = pdata.diagnosis                          # M or B 
mylist = ['_c32','id','diagnosis']
x = pdata.drop(mylist,axis = 1 )
import time
sns.set(style="whitegrid", palette="muted")
data_dia = y
data = x
data_n_2 = (data - data.mean()) / (data.std())              # standardization
data = pd.concat([y,data_n_2.iloc[:,0:10]],axis=1)
data = pd.melt(data,id_vars="diagnosis",
                    var_name="features",
                    value_name='value')
fig=plt.figure(figsize=(10,10))
tic = time.time()
sns.swarmplot(x="features", y="value", hue="diagnosis", data=data)
display(fig)
plt.xticks(rotation=90)

In [30]:
data = pd.concat([y,data_n_2.iloc[:,10:20]],axis=1)
data = pd.melt(data,id_vars="diagnosis",
                    var_name="features",
                    value_name='value')
fig= plt.figure(figsize=(10,10))
sns.swarmplot(x="features", y="value", hue="diagnosis", data=data)
plt.xticks(rotation=90)
display(fig)

In [31]:
data = pd.concat([y,data_n_2.iloc[:,20:31]],axis=1)
data = pd.melt(data,id_vars="diagnosis",
                    var_name="features",
                    value_name='value')
fig=plt.figure(figsize=(10,10))
sns.swarmplot(x="features", y="value", hue="diagnosis", data=data)
toc = time.time()
plt.xticks(rotation=90)
print("swarm plot time: ", toc-tic ," s")
display(fig)

These swarmplots help to see variance more clearly. In these three plots  feature which look like more clear in terms of classification are area_worst which in the last swarm plot looks like malignant and benign are seprated not totaly but mostly. Hovewer, smoothness_se in swarm plot 2 looks like malignant and benign are mixed.

In [33]:
from mpl_toolkits import mplot3d
import numpy as np
import matplotlib.pyplot as plt
fig = plt.figure()
ax = plt.axes(projection="3d")
z_points = pdata['radius_mean']
x_points = pdata['area_mean']
y_points = pdata['perimeter_mean']
ax.scatter3D(x_points, y_points, z_points, c=z_points, cmap='hsv');
display(fig)

This 3d scatter plot helps us to visualize how closely radius_mean, area_mean and perimeter_mean are to each other as we had seen earlier through the heatmap and we can see that they are closely related.

In [35]:
from mpl_toolkits import mplot3d
import numpy as np
import matplotlib.pyplot as plt
fig = plt.figure()
ax = plt.axes(projection="3d")
z_points = pdata['compactness_mean']
x_points = pdata['area_se']
y_points = pdata['fractal_dimension_worst']
ax.scatter3D(x_points, y_points, z_points, c=z_points, cmap='hsv');
display(fig)

This 3d scatter plot helps us to visualize how closely compactness_mean, area_se and fractal_dimension_worst are to each other as we had seen earlier through the heatmap and we can see that the correlation between these 3 values is not as strong.

## Data Cleaning
Only '_c32' is completely null, probably an error in the dataset, we drop it in below and we drop null values

In [38]:
df=df.drop('_c32')

In [39]:
df.printSchema()

root
-- id: integer (nullable = true)
-- diagnosis: string (nullable = true)
-- radius_mean: double (nullable = true)
-- texture_mean: double (nullable = true)
-- perimeter_mean: double (nullable = true)
-- area_mean: double (nullable = true)
-- smoothness_mean: double (nullable = true)
-- compactness_mean: double (nullable = true)
-- concavity_mean: double (nullable = true)
-- concave points_mean: double (nullable = true)
-- symmetry_mean: double (nullable = true)
-- fractal_dimension_mean: double (nullable = true)
-- radius_se: double (nullable = true)
-- texture_se: double (nullable = true)
-- perimeter_se: double (nullable = true)
-- area_se: double (nullable = true)
-- smoothness_se: double (nullable = true)
-- compactness_se: double (nullable = true)
-- concavity_se: double (nullable = true)
-- concave points_se: double (nullable = true)
-- symmetry_se: double (nullable = true)
-- fractal_dimension_se: double (nullable = true)
-- radius_worst: double (nullable = true)
-- texture_worst: double (nullable = true)
-- perimeter_worst: double (nullable = true)
-- area_worst: double (nullable = true)
-- smoothness_worst: double (nullable = true)
-- compactness_worst: double (nullable = true)
-- concavity_worst: double (nullable = true)
-- concave points_worst: double (nullable = true)
-- symmetry_worst: double (nullable = true)
-- fractal_dimension_worst: double (nullable = true)

## Data Transformation and modeling

In [41]:
splitSample = df.randomSplit([.70,.30], 123)
trainingData = splitSample[0]
testData = splitSample[1]

##To predict the Diagnosis column(which shows if the tumor is malignant or benign) we will use the following columns as determinant values or features:
* texture_mean
* perimeter_mean
* smoothness_mean
* area_mean
* radius_mean
* compactness_mean
* concavity_mean
* concave points_mean
* symmetry_mean
* fractal_dimension_mean

In [43]:
from pyspark.ml import *
from pyspark.ml.feature import *
from pyspark.ml.feature import Bucketizer
from pyspark.ml.classification import LogisticRegression,RandomForestClassifier, NaiveBayes
from pyspark.ml.tuning import *
from pyspark.ml.evaluation import *
from pyspark.ml.regression import *
from pyspark.ml.feature import StringIndexer, IndexToString, VectorAssembler
from pyspark.ml.evaluation import RegressionEvaluator 
from pyspark.ml import Pipeline, Model, PipelineModel


stringIndexer_label = StringIndexer(inputCol="diagnosis", outputCol="label").fit(df) 
#Note the above outputCol is label (the predicted column). Here we predict the diagnosis from the attributes below.

vectorAssembler_features = VectorAssembler(inputCols=["texture_mean","perimeter_mean", "smoothness_mean", "area_mean","radius_mean","compactness_mean","concavity_mean","concave points_mean","symmetry_mean","fractal_dimension_mean"], outputCol="features")



pipeline = Pipeline(stages=[])
basePipeline =[stringIndexer_label,vectorAssembler_features] #creating a basepipeline


#Modeling using Logistic Regression
lr = LogisticRegression(maxIter=10)
pl_lr = basePipeline + [lr]
pg_lr = ParamGridBuilder()\
          .baseOn({pipeline.stages: pl_lr})\
          .addGrid(lr.regParam,[0.01, .04])\
          .addGrid(lr.elasticNetParam,[0.1, 0.4])\
          .build()


rf = RandomForestClassifier(numTrees=50)
pl_rf = basePipeline + [rf]
pg_rf = ParamGridBuilder()\
      .baseOn({pipeline.stages: pl_rf})\
      .build()

nb = NaiveBayes()
pl_nb = basePipeline + [nb]
pg_nb = ParamGridBuilder()\
      .baseOn({pipeline.stages: pl_nb})\
      .addGrid(nb.smoothing,[0.4,1.0])\
      .build()

# Create the stages pipeline with all the feature transformers to create an Estimator
paramGrid = pg_lr + pg_rf + pg_nb


In [44]:
model = pipeline.fit(trainingData)

In [45]:
%sh
pip install mlflow            

Collecting mlflow
Requirement already satisfied: pandas in /databricks/python3/lib/python3.7/site-packages (from mlflow) (0.24.2)
Collecting querystring-parser (from mlflow)
Collecting protobuf>=3.6.0 (from mlflow)
 Using cached https://files.pythonhosted.org/packages/c8/bd/9609c681f655c4b85f4e4f4c99d42b68bca63ad8891c6924ba6696dee0bb/protobuf-3.11.1-cp37-cp37m-manylinux1_x86_64.whl
Collecting simplejson (from mlflow)
Collecting sqlalchemy (from mlflow)
Requirement already satisfied: python-dateutil in /databricks/python3/lib/python3.7/site-packages (from mlflow) (2.8.0)
Collecting gitpython>=2.1.0 (from mlflow)
 Using cached https://files.pythonhosted.org/packages/20/8c/4543981439d23c4ff65b2e62dddd767ebc84a8e664a9b67e840d1e2730d3/GitPython-3.0.5-py3-none-any.whl
Collecting Flask (from mlflow)
 Using cached https://files.pythonhosted.org/packages/9b/93/628509b8d5dc749656a9641f4caf13540e2cdec85276964ff8f43bbb1d3b/Flask-1.1.1-py2.py3-none-any.whl
Collecting docker>=4.0.0 (from mlflow)
 Using cached https://files.pythonhosted.org/packages/cc/ca/699d4754a932787ef353a157ada74efd1ceb6d1fc0bfb7989ae1e7b33111/docker-4.1.0-py2.py3-none-any.whl
Collecting gorilla (from mlflow)
 Using cached https://files.pythonhosted.org/packages/e3/56/5a683944cbfc77e429c6f03c636ca50504a785f60ffae91ddd7f5f7bb520/gorilla-0.3.0-py2.py3-none-any.whl
Requirement already satisfied: six>=1.10.0 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.12.0)
Collecting pyyaml (from mlflow)
Collecting click>=7.0 (from mlflow)
 Using cached https://files.pythonhosted.org/packages/fa/37/45185cb5abbc30d7257104c434fe0b07e5a195a6847506c074527aa599ec/Click-7.0-py2.py3-none-any.whl
Collecting databricks-cli>=0.8.7 (from mlflow)
Collecting sqlparse (from mlflow)
 Using cached https://files.pythonhosted.org/packages/ef/53/900f7d2a54557c6a37886585a91336520e5539e3ae2423ff1102daf4f3a7/sqlparse-0.3.0-py2.py3-none-any.whl
Collecting gunicorn; platform_system != "Windows" (from mlflow)
 Using cached https://files.pythonhosted.org/packages/69/ca/926f7cd3a2014b16870086b2d0fdc84a9e49473c68a8dff8b57f7c156f43/gunicorn-20.0.4-py2.py3-none-any.whl
Requirement already satisfied: numpy in /databricks/python3/lib/python3.7/site-packages (from mlflow) (1.16.2)
Collecting alembic (from mlflow)
Collecting entrypoints (from mlflow)
 Using cached https://files.pythonhosted.org/packages/ac/c6/44694103f8c221443ee6b0041f69e2740d89a25641e62fb4f2ee568f2f9c/entrypoints-0.3-py2.py3-none-any.whl
Collecting cloudpickle (from mlflow)
 Using cached https://files.pythonhosted.org/packages/c1/49/334e279caa3231255725c8e860fa93e72083567625573421db8875846c14/cloudpickle-1.2.2-py2.py3-none-any.whl
Requirement already satisfied: requests>=2.17.3 in /databricks/python3/lib/python3.7/site-packages (from mlflow) (2.21.0)
Requirement already satisfied: pytz>=2011k in /databricks/python3/lib/python3.7/site-packages (from pandas->mlflow) (2018.9)
Requirement already satisfied: setuptools in /usr/lib/python3.7/site-packages (from protobuf>=3.6.0->mlflow) (40.8.0)
Collecting gitdb2>=2.0.0 (from gitpython>=2.1.0->mlflow)
 Using cached https://files.pythonhosted.org/packages/03/6c/99296f89bad2ef85626e1df9f677acbee8885bb043ad82ad3ed4746d2325/gitdb2-2.0.6-py2.py3-none-any.whl
Collecting Jinja2>=2.10.1 (from Flask->mlflow)
 Using cached https://files.pythonhosted.org/packages/65/e0/eb35e762802015cab1ccee04e8a277b03f1d8e53da3ec3106882ec42558b/Jinja2-2.10.3-py2.py3-none-any.whl
Collecting Werkzeug>=0.15 (from Flask->mlflow)
 Using cached https://files.pythonhosted.org/packages/ce/42/3aeda98f96e85fd26180534d36570e4d18108d62ae36f87694b476b83d6f/Werkzeug-0.16.0-py2.py3-none-any.whl
Collecting itsdangerous>=0.24 (from Flask->mlflow)
 Using cached https://files.pythonhosted.org/packages/76/ae/44b03b253d6fade317f32c24d100b3b35c2239807046a4c953c7b89fa49e/itsdangerous-1.1.0-py2.py3-none-any.whl
Collecting websocket-client>=0.32.0 (from docker>=4.0.0->mlflow)
 Using cached https://files.pythonhosted.org/packages/29/19/4475

In [46]:
from pyspark.ml import Pipeline, PipelineModel
from pyspark.ml.feature import Bucketizer, RegexTokenizer, StopWordsRemover, CountVectorizer, IDF, NGram, StringIndexer, VectorAssembler
from pyspark.ml.classification import LogisticRegression, RandomForestClassifier, GBTClassifier, NaiveBayes
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder
from pyspark.ml.evaluation import BinaryClassificationEvaluator
# 20 minutes with 4 models and data limited to 2000
# 10 minutes with 4 models and data limited to 1000
# 2.5 minutes with 4 models and data limited to 100

# Should run more than 2 folds, but here we simplify so that it will complete
# Evaluate with BinaryClassificationEvaluator because there are only two labels (associated with the Buckets defined above)
cv = CrossValidator()\
      .setEstimator(pipeline)\
      .setEvaluator(BinaryClassificationEvaluator())\
      .setEstimatorParamMaps(paramGrid)\
      .setNumFolds(2)

cvModel = cv.fit(trainingData.limit(500))
#cvModel = cv.fit(training.limit(1000))
#cvModel = cv.fit(df) # more than 1 hour for all data (if completes)

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [47]:
import numpy as np
# BinaryClassificationEvaluator defaults to ROC AUC, so higher is better
# http://gim.unmc.edu/dxtests/roc3.htm
cvModel.getEstimatorParamMaps()[ np.argmax(cvModel.avgMetrics) ]

Out[67]: {Param(parent='Pipeline_6579b7fe5e22', name='stages', doc='a list of pipeline stages'): [StringIndexer_452bf685ac1d,
 VectorAssembler_dd5bbbc8c074,
 RandomForestClassifier_989b9e2cab7b]}

In [48]:
cvModel.getEstimatorParamMaps()[ np.argmin(cvModel.avgMetrics) ]

Out[68]: {Param(parent='Pipeline_6579b7fe5e22', name='stages', doc='a list of pipeline stages'): [StringIndexer_452bf685ac1d,
 VectorAssembler_dd5bbbc8c074,
 NaiveBayes_53378efb4a1b],
 Param(parent='NaiveBayes_53378efb4a1b', name='smoothing', doc='The smoothing parameter, should be >= 0, default is 1.0'): 0.4}

In [49]:
from pyspark.ml.feature import StandardScaler
from pyspark.ml.linalg import Vectors
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.clustering import KMeans
import re
def paramGrid_model_name(model):
  params = [v for v in model.values() if type(v) is not list]
  name = [v[-1] for v in model.values() if type(v) is list][0]
  name = re.match(r'([a-zA-Z]*)', str(name)).groups()[0]
  return "{}{}".format(name,params)

# Resulting metric and model description
# get the measure from the CrossValidator, cvModel.avgMetrics
# get the model name & params from the paramGrid
# put them together here:
kmeans_measures = zip(cvModel.avgMetrics, [paramGrid_model_name(model) for model in paramGrid])
metrics,model_names = zip(*kmeans_measures)

In [50]:
import seaborn as sns
import matplotlib.pyplot as plt

plt.clf() # clear figure
fig = plt.figure( figsize=(5, 5))
plt.style.use('fivethirtyeight')
axis = fig.add_axes([0.1, 0.3, 0.8, 0.6])
# plot the metrics as Y
#plt.plot(range(len(model_names)),metrics)
plt.bar(range(len(model_names)),metrics)
# plot the model name & param as X labels
plt.xticks(range(len(model_names)), model_names, rotation=70, fontsize=6)
plt.yticks(fontsize=6)
#plt.xlabel('model',fontsize=8)
plt.ylabel('ROC AUC (better is greater)',fontsize=8)
plt.title('Model evaluations')
display(plt.show())

In [51]:
rfBest = cvModel.bestModel
# Model quality
predictions = rfBest.transform(testData)
evaluatorRF = MulticlassClassificationEvaluator(labelCol="label", predictionCol="prediction", metricName="accuracy")
accuracy = evaluatorRF.evaluate(predictions)
print("Accuracy = %g" % accuracy)
print("Test Error = %g" % (1.0 - accuracy))

Accuracy = 0.944134
Test Error = 0.0558659

In [52]:
predictions = cvModel.transform(testData)
display(predictions.select('label','prediction','texture_mean','perimeter_mean', 'smoothness_mean', 'area_mean','radius_mean','compactness_mean','concavity_mean','concave points_mean','symmetry_mean','fractal_dimension_mean').limit(100))

label prediction texture_mean perimeter_mean smoothness_mean area_mean radius_mean compactness_mean concavity_mean concave points_mean symmetry_mean fractal_dimension_mean 0.0 0.0 13.12 81.89 0.06955 515.9 12.89 0.03729 0.0226 0.01171 0.1337 0.05581 0.0 0.0 19.1 97.03 0.08992 687.3 14.96 0.09823 0.0594 0.04819 0.1879 0.05852 0.0 0.0 16.17 83.18 0.09879 507.6 12.94 0.08836 0.03296 0.0239 0.1735 0.062 0.0 1.0 20.76 82.15 0.09933 480.4 12.63 0.1209 0.1065 0.06021 0.1735 0.0707 1.0 1.0 17.12 121.4 0.1054 1077.0 18.66 0.11 0.1457 0.08665 0.1966 0.06213 1.0 1.0 11.89 102.5 0.1257 736.9 15.46 0.1555 0.2032 0.1097 0.1966 0.07069 1.0 1.0 23.75 91.56 0.1323 597.8 13.81 0.1768 0.1558 0.09176 0.2251 0.07421 0.0 0.0 18.6 81.09 0.09965 481.9 12.47 0.1058 0.08005 0.03821 0.1925 0.06373 0.0 0.0 16.54 94.15 0.09751 641.2 14.42 0.1139 0.08007 0.04223 0.1912 0.06412 0.0 0.0 22.72 78.75 0.06935 447.8 12.05 0.1073 0.07943 0.02978 0.1203 0.06659 1.0 1.0 23.29 93.97 0.08682 664.7 14.6 0.06636 0.0839 0.05271 0.1627 0.05416 0.0 0.0 19.24 88.73 0.07991 602.9 13.9 0.05326 0.02995 0.0207 0.1579 0.05594 0.0 0.0 16.32 81.25 0.1158 476.3 12.54 0.1085 0.05928 0.03279 0.1943 0.06612 0.0 0.0 13.1 54.53 0.1036 221.3 8.571 0.07632 0.02565 0.0151 0.1678 0.07126 0.0 0.0 24.54 47.92 0.05263 181.0 7.76 0.04362 0.0 0.0 0.1587 0.05884 1.0 1.0 17.77 132.9 0.08474 1326.0 20.57 0.07864 0.0869 0.07017 0.1812 0.05667 1.0 1.0 21.82 87.5 0.1273 519.8 13.0 0.1932 0.1859 0.09353 0.235 0.07389 1.0 1.0 24.8 132.4 0.0974 1123.0 19.17 0.2458 0.2065 0.1118 0.2397 0.078 1.0 1.0 23.95 103.7 0.08401 782.7 15.85 0.1002 0.09938 0.05364 0.1847 0.05338 1.0 0.0 20.13 94.74 0.09867 684.5 14.68 0.072 0.07395 0.05259 0.1586 0.05922 1.0 1.0 23.04 137.2 0.09428 1404.0 21.16 0.1022 0.1097 0.08632 0.1769 0.05278 1.0 1.0 15.05 115.0 0.09847 955.1 17.57 0.1157 0.09875 0.07953 0.1739 0.06149 1.0 1.0 25.11 124.8 0.1064 1088.0 18.63 0.1887 0.2319 0.1244 0.2183 0.06197 1.0 1.0 17.88 107.0 0.104 807.2 16.13 0.1559 0.1354 0.07752 0.1998 0.06515 0.0 0.0 18.42 82.61 0.08983 523.8 13.03 0.03766 0.02562 0.02923 0.1467 0.05863 1.0 0.0 25.2 95.54 0.09387 698.8 14.99 0.05131 0.02398 0.02899 0.1565 0.05504 1.0 0.0 21.58 86.18 0.08162 563.0 13.44 0.06031 0.0311 0.02031 0.1784 0.05587 1.0 1.0 20.28 87.32 0.1041 545.2 13.28 0.1436 0.09847 0.06158 0.1974 0.06782 1.0 1.0 18.7 120.3 0.1148 1033.0 18.22 0.1485 0.1772 0.106 0.2092 0.0631 1.0 0.0 22.02 97.26 0.09056 712.8 15.1 0.07081 0.05253 0.03334 0.1616 0.05684 1.0 1.0 21.59 95.55 0.1137 656.9 14.71 0.1365 0.1293 0.08123 0.2027 0.06758 0.0 0.0 19.31 82.61 0.0806 527.2 13.05 0.03789 6.92E-4 0.004167 0.1819 0.05501 1.0 1.0 22.15 96.42 0.1049 645.7 14.25 0.2008 0.2135 0.08653 0.1949 0.07292 0.0 0.0 20.97 73.81 0.1102 401.5 11.45 0.09362 0.04591 0.02233 0.1842 0.07005 0.0 0.0 18.3 86.6 0.1022 555.1 13.45 0.08165 0.03974 0.0278 0.1638 0.0571 1.0 1.0 15.56 87.21 0.1398 530.2 13.11 0.1765 0.2071 0.09601 0.1925 0.07692 0.0 0.0 16.99 62.5 0.1037 290.2 9.777 0.08404 0.04334 0.01778 0.1584 0.07065 0.0 0.0 15.83 55.84 0.115 230.9 8.726 0.08201 0.04132 0.01924 0.1649 0.07633 0.0 0.0 15.76 58.74 0.09462 245.2 8.95 0.1243 0.09263 0.02308 0.1305 0.07163 1.0 1.0 22.91 105.7 0.1155 782.6 15.78 0.1752 0.2133 0.09479 0.2096 0.07331 0.0 0.0 10.82 73.34 0.09373 403.3 11.41 0.06685 0.03512 0.02623 0.1667 0.06113 0.0 0.0 10.89 94.28 0.1101 640.7 14.5 0.1099 0.08842 0.05778 0.1856 0.06402 0.0 0.0 16.39 86.1 0.07115 553.5 13.37 0.07325 0.08092 0.028 0.1422 0.05823 1.0 1.0 21.91 120.2 0.0943 1075.0 18.45 0.09709 0.1153 0.06847 0.1692 0.05727 1.0 0.0 22.47 81.72 0.09055 506.3 12.77 0.05761 0.04711 0.02704 0.1585 0.06065 0.0 0.0 16.67 74.72 0.1051 423.6 11.71 0.06095 0.03592 0.026 0.1339 0.05945 1.0 1.0 17.57 96.85 0.1167 678.1 14.95 0.1305 0.1539 0.08624 0.1957 0.06216 0.0 0.0 17.31 73.66 0.1092 398.0 11.43 0.09486 0.02031 0.01861 0.1645 0.06562 1.0 1.0 16.58 78.99 0.1091 432.0 11.8 0.17 0.1659 0.07415 0.2678 0.07371 0.0 0.0 17.91 88.12 0.07944 585.0 13.74 0.06376 0.02881 0.01329 0.1473 0.0558 

##Evaluating the Model and Visualizing

In [54]:
correct = predictions.where("(label = prediction)").count()
incorrect = predictions.where("(label != prediction)").count()

resultDF = sqlContext.createDataFrame([['correct', correct], ['incorrect', incorrect]], ['metric', 'value'])
display(resultDF)

metric value correct 169 incorrect 10

In [55]:
counts = [predictions.where('label=1').count(), predictions.where('prediction=1').count(),
          predictions.where('label=0').count(), predictions.where('prediction=0').count()]
names = ['actual 1', 'predicted 1', 'actual 0', 'predicted 0']
display(sqlContext.createDataFrame(zip(names,counts),['Measure','Value']))

Measure Value actual 1 73 predicted 1 67 actual 0 106 predicted 0 112

In [56]:
from pyspark.mllib.evaluation import BinaryClassificationMetrics
# https://spark.apache.org/docs/latest/mllib-evaluation-metrics.html
# Compute raw scores on the test set
predictionAndLabels = predictions.select('prediction','label').rdd

# Instantiate metrics object
metrics = BinaryClassificationMetrics(predictionAndLabels)

# Area under precision-recall curve
print("Area under PR = %s" % metrics.areaUnderPR)

# Area under ROC curve
print("Area under ROC = %s" % metrics.areaUnderROC)

Area under PR = 0.9393367592282713
Area under ROC = 0.9357715171879039

##RESULT:
* We can see that the Random Forest Classifier Model performed the best with an accuracy of 94.41% and the worst model was Naive Bayes Classifier. 
* Random Forest Classifier correctly identified 169 instances out of the test data set and didn't correctly identify 10 instances. 
* It correctly predicted 73 instances of label 1 which were actually 1 and incorrectly identified 67 instances which were not actually 1.
* Similarly it correctly predicted 106 instances of label 0 which were actually 0 and incorrectly identified 112 instances which were not actually 0.
* Thus we can conclude Random Forest is the best for the prediction of this cancer dataset.

##SPARK API's we missed out on:

* Graphx[3] is a spark API for graph and graph-parallel computation. Graph algorithms are iterative in nature and properties of vertices depends upon the properties of its directly or indirectly (connected via other vertices) connected vertices.
* Pregel is a vertex centric graph processing model developed by Google and spark graphx provides an optimized variant of pregel api.
* Unlike the original Pregel API, the GraphX Pregel API factors the sendMessage computation over edges, enables the message sending computation to read both vertex attributes, and constrains messages to the graph structure. These changes allow for substantially more efficient distributed execution while also exposing greater flexibility for graph-based computation.

Pregel API processing consists of executing super steps:

Superstep 0:
* Pass initial message to all the vertices
* Send the value as message to its directly connected vertices

Superstep 1:
* Receive messages from the previous steps
* Mutate the value
* Send the value as message to its directly connected vertices
* Repeat Superstep 1 until there are messages to pass and stop when there are no more messages to be passed.

In [61]:
displayHTML("<img src ='http://artimg.ishenping.com/20190617172739113_BSSFET.jpg'>")

In [62]:
displayHTML("<img src ='http://artimg.ishenping.com/20190617172752565_XJFTIF.jpg'>")

In [63]:
displayHTML("<img src ='http://artimg.ishenping.com/20190617172801635_ISUDTZ.jpg'>")

In [64]:
displayHTML("<img src ='http://artimg.ishenping.com/20190617172809895_HZDFZI.jpg'>")